### DISCUSSION TITLE
Provide an array of interfaces as a dependency

### DISCUSSION THREAD

**Discussion by NicklasWallgren** [uber-go/fx#880] [Q&A] [answered]
Hey, 

I'm trying to setup an array of interfaces (`[]HealthChecker`) as a dependency to my `HealthCheckerManager`. 
What is the proper way of setting this up with fx?

Thanks.

```
type HealthChecker interface{}

type DBHealthChecker struct {}
func NewDBHealthChecker(....) HealthChecker {
	return &DBHealthChecker{....}
}

type HealthCheckerManager struct{
  healthCheckers []HealthChecker
}
func NewHealthCheckerManager(healthCheckers []HealthChecker) *HealthCheckerManager {
	return &HealthCheckerManager{healthCheckers: healthCheckers}
}
```

What I've tried with so far.
```
var dbChecker = fx.Provide(fx.Annotated{
	Name:   "dbChecker",
	Target: health.NewDBHealthChecker,
})

type Params struct {
	fx.In
	DBChecker health.HealthChecker `name:"dbChecker"`
}

var healthCheckers = fx.Provide(func(healthCheckers Params) []health.HealthChecker {
	return []health.HealthChecker{healthCheckers.DBChecker}
})

var healthCheckerManager = fx.Provide(health.NewHealthCheckerManager)
```

**Comment by abhinav** [comment, marked as answer]
Hey there! It looks like what you're looking for is [value groups](https://pkg.go.dev/go.uber.org/fx#hdr-Value_Groups).

Value groups are a feature of Fx that allow:
- producers to produce singular values to a named group
- consumers to consume all instances of those values produces from all producers in a single slice

You can use value groups in one of two forms:

- Parameter and result structs with [fx.In](https://pkg.go.dev/go.uber.org/fx#In) and [fx.Out](https://pkg.go.dev/go.uber.org/fx#Out)
- Annotations with [fx.Annotate](https://pkg.go.dev/go.uber.org/fx#Annotate)

## Value groups with fx.In and fx.Out

To **produce** values to a value group, make the constructor for your producer return a struct that embeds fx.Out, and annotate the field for the value with `group:"name"` where `"name"` is the name of your value group.

```go
type Result struct {
  fx.Out

  Result HealthChecker `group:"checkers"`
}
 
func NewDBHealthChecker(...) Result {
  return ...
}
```

To **consume** a value group, have your constructor accept a struct that embeds fx.In, and annotate the field with `group:"name"` using the same group name as before. This field should be a slice of values of the type returned by the producer.

```go
type Params struct {
  fx.In

  Checkers []HealthChecker `group:"checkers"`
}

func NewHealthCheckManager(p Params) HealthCheckManager {
  // ...
}
```

## Value groups with annotations

A more concise and conveient form of using value groups is with `fx.Annotate`.

To **produce** values to a value group, leave your original constructor unchanged and annotate it when you provide it to Fx—either via [`fx.Module`](https://pkg.go.dev/go.uber.org/fx#Module) or directly to [`fx.New`](https://pkg.go.dev/go.uber.org/fx#New)—with [fx.ResultTags](https://pkg.go.dev/go.uber.org/fx#ResultTags) using the `group:"name"` annotation.

```go
func NewDBHealthChecker(...) HealthChecker { ... }

// Elsewhere,

fx.Provide(
  fx.Annotate(NewDBHealthChecker, fx.ResultTags(`group:"checkers"`)),
)

// Or,

var Module = fx.Module("DB",
  fx.Provide(
    fx.Annotate(NewDBHealthChecker, fx.ResultTags(`group:"checkers"`)),
    // other provides for the DB module
  ),
)
```

> 💡 **TIP**: If `NewDBHealthChecker` returns a `*DBHealthChecker` instead of `HealthChecker`, you don't need to change it. You can combine it fx.ResultTags with [fx.As](https://pkg.go.dev/go.uber.org/fx#As) to turn it into a HealthChecker interface and provide it to the value group like so:
>
> ```go
> fx.Annotate(NewDBHealthChecker, fx.As(new(HealthChecker)), fx.ResultTags(`group:"checkers"`))
> ```

To **consume** a value group, leave your original constructor similarly unchanged, provided that it accepts a slice of `HealthChecker`s, and annotate it with [fx.ParamTags](https://pkg.go.dev/go.uber.org/fx#ParamTags) when you provide it to Fx.

```go
func NewHealthCheckerManager(healthCheckers []HealthChecker) *HealthCheckerManager {
  ...
}

// Elsewhere,

fx.Provide(
  fx.Annotate(NewHealthCheckerManager, fx.ParamTags(`group:"checkers"`)),
)
```

If your function accepts other parameters, you can pass `""` to fx.ParamTags for those:

```go
// Given,
// func NewHealthCheckerManager(cfg *Config, checkers []HealthChecker) *HealthCheckManager

fx.Annotate(NewHealthCheckerManager, fx.ParamTags("", `group:"checkers"`)),
```

Hope this helps!

**Reply by NicklasWallgren** [reply]
Thanks!

**Reply by NicklasWallgren** [reply]
@abhinav Is it possible to order the values associated with a group-tag? Either by index, weight or the order which they have been registered?

It seems as the order of the values differs on each application start.

```
fx.Provide(
  fx.Annotate(NewDBHealthChecker, fx.ResultTags(`group:"checkers"`)),  // <--- Should always have index 0
  fx.Annotate(NewRabbitMQHealthChecker, fx.ResultTags(`group:"checkers"`)), // <--- Should always have index 1
  fx.Annotate(NewHttpHealthChecker, fx.ResultTags(`group:"checkers"`)), // <--- Should always have index 2
  ....
  fx.Annotate(NewHealthCheckerManager, fx.ParamTags("", `group:"checkers"`)),
)
```

**Reply by NicklasWallgren** [reply]
Ah, it has already been asked before. https://github.com/uber-go/fx/issues/807
